### ISSUE TITLE
Profiles not loading on external pages

### ISSUE THREAD

**Issue by netzih** [progressivetech/net.ourpowerbase.remoteform#12]
I added the URL's to the remote forms settings and enabled remote forms in the profile.
When putting it on the external website I'm getting "Access to XMLHttpRequest at 'https://www.example.com/wp-admin/admin.php?page=CiviCRM&amp;q=civicrm%2Fremoteform' from origin 'https://www.remoteexample.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request."

This is on Wordpress.

**Comment by jmcclelland** [comment]
Can you double check to make sure that:  https://www.remoteexample.com is entered on the remote forms settings page (if you access the page from both www.remoteexample.com and remoteexample.com then both should be enetered on the page).

It should be entered as:

```
https://www.remoteexample.com
https://remoteexample.com
```

**Comment by netzih** [comment]
Yes, I added them both.
Is this maybe a WordPress issue?
Was this tested on WordPress?

**Comment by jmcclelland** [comment]
It might be a wordpress issue (I do all testing on Drupal). However, the error doesn't sound like it would be wordpress specific.

I noticed that the error says: Redirect is not allowed for a pre-flight test.

Is it possible that the CiviCRM URL in your javascript code is generating a redirect? The CiviCRM url should be the final version of the URL that is used when accessing your database. In other words, http://remoteexample.com might get redirected to https://remoteexample.com or https://www.remoteexample.com might get redirected to https://remoteexample.com.

Also - here is a useful post for testing CORS headers to make sure they are set properly in CiviCRM.

https://stackoverflow.com/questions/12173990/how-can-you-debug-a-cors-request-with-curl

**Comment by netzih** [comment]
No, it is always https://www.remoteexample.com
I copied it directly from the console just to make sure it was the same.
And it was all correct,
Is there possibly some reason the remote forms settings isn't writing the
url to civicrm cors policy?

**Comment by jmcclelland** [comment]
Here's a command that demonstrates CORS is working properly on our database (note, this is specific to our web site, our civi database and a Drupal URL):

```
 curl -H "Origin: https://progressivetech.org" -H "Access-Control-Request-Method: POST" -H "Access-Control-Request-Headers: X-Requested-With" -X OPTIONS --verbose https://ptp.ourpowerbase.net/civicrm/remoteform 2>&1 | grep Allow-Origin
```

It outputs:

```
< Access-Control-Allow-Origin: https://progressivetech.org
```

If you have access to a Linux or Mac machine - you should be able to run that command substituting the right values for your setup. Then we can pinpoint the problem to setting CORS headers (or something else if they are set properly).

**Comment by netzih** [comment]
Tried putting it in. Not getting the same output.
seems it's not setting the cors headers properly.

**Comment by gah242s** [comment]
I feel like that was the same issue that I hit with issue #10 though it's been long enough now that I may be incorrect.  However, I know that I did have a CORS error, and probably just failed to put that in that bug report.  Might be worth trying the workaround and see if that clears it up.  And if it does, I apologize I didn't put the CORS error in the report.

**Comment by netzih** [comment]
Changed the URL as you suggested.
Now the error is a bit different  in console
"Cross-Origin Request Blocked: The Same Origin Policy disallows reading the
remote resource at
https://www.chabadofrichmond.com/?page=CiviCRM&amp;q=civicrm%2Fremoteform.
(Reason: CORS header ‘Access-Control-Allow-Origin’ missing)"
So it seems that remoteforms setting is not adding the allow access control
header.

Is there another way to add the header manually?

**Comment by gah242s** [comment]
Ugh.  Looks like I have bad documentation there.  I think, in the .htaccess file, I did something like this:
```
    <IfModule mod_headers.c>
        SetEnvIf Origin "http(s)?://(www\.)?(domain1.org|domain2.com)$" AccessControlAllowOrigin=$0
        Header add Access-Control-Allow-Origin %{AccessControlAllowOrigin}e env=AccessControlAllowOrigin
    </IfModule>
```

**Comment by netzih** [comment]
Thank you so much for that.
Looks like we're getting somewhere.
I added that to my .htaccess file and now the error I'm getting is
" Response to preflight request doesn't pass access control check: The
'Access-Control-Allow-Origin' header contains multiple values '
https://www.domain.com, https://www.domain.com', but only one is allowed."
I removed "|domain2.com" from what you wrote so it just shows "(domain1.org)"
and that is what it says. Is there something else in the code that is
adding it twice?

**Comment by gah242s** [comment]
Doh!  Should have mentioned that domain1 and domain2 were just placeholders for the actual value of the site sending the request, so whatever remote sites where the form is viewed from.  So, from your examples above, it should be:
domain1.org -> www.remoteexample.com
domain2.com -> remoteexample.com

**Comment by netzih** [comment]
Not understanding.
I want the form hosted on https://www.example1.com to work on
https://www.remoteexample.com (only need it working on one external domain
for now)
What do I need to put there?

Right now I have it like this <IfModule mod_headers.c>

        SetEnvIf Origin "http(s)?://(www\.)?(remoteexample.com)$"
AccessControlAllowOrigin=$0
        Header add Access-Control-Allow-Origin
%{AccessControlAllowOrigin}e env=AccessControlAllowOrigin
    </IfModule>

**Comment by netzih** [comment]
The error I'm getting now is "Request header field content-type is not
allowed by Access-Control-Allow-Headers in preflight response."

**Comment by gah242s** [comment]
Yes, the .htaccess code looks like what I would expect.  I don't recall seeing that latest message about the content-type, so I'm not sure I'm much more help.  The browser dev tools inspector should say what kind of content type and might result in some better Googling about.

**Comment by netzih** [comment]
Played around a bit and based on what I found on google changed the
.htaccess file to say
   <IfModule mod_headers.c>
    Header set Access-Control-Allow-Origin "https://www.remoteexample.com"
    Header set Access-Control-Allow-Methods "POST, GET, OPTIONS, DELETE,
PUT"
    Header set Access-Control-Max-Age "1000"
    Header set Access-Control-Allow-Headers "x-requested-with,
Content-Type, Ac$

</IfModule>

I also changed the URL within the HTML code as you said in the post
(removing the wp-admin part)

Now in console I'm getting a different error "SyntaxError: "JSON.parse:
unexpected character at line 1 column 1 of the JSON data"

   - onreadystatechange
   https://www.chabadofrichmond.com/wp-content/uploads/civicrm/ext/net.ourpowerbase.remoteform-master/remoteform.js
   :647"

**Comment by jmcclelland** [comment]
Hi all - sorry for the slow responses. I finally had a chance to install wordpress and test and I found two errors in my code:

1. The code generated was giving you a backend URL instead of a frontend URL. I thought I had fixed that, but it turns out I made a mistake in my code and was missing a variable.
2. The code generated a fully url encoded post address, which caused problems.

Both problems should be fixed with the most recent version. Note: you should not need any changes to any .htaccess files, just the fixed remote form code (which should give you a new post url).

Can you give it a shot and let me know if it does the trick?

**Comment by netzih** [comment]
> Hi all - sorry for the slow responses. I finally had a chance to install wordpress and test and I found two errors in my code:
> 
> 1. The code generated was giving you a backend URL instead of a frontend URL. I thought I had fixed that, but it turns out I made a mistake in my code and was missing a variable.
> 2. The code generated a fully url encoded post address, which caused problems.
> 
> Both problems should be fixed with the most recent version. Note: you should not need any changes to any .htaccess files, just the fixed remote form code (which should give you a new post url).
> 
> Can you give it a shot and let me know if it does the trick?

Awesome!
Just tested it for profiles and it worked,
On contributions seems to be working too but did not want to test any cards just yet...
Can't wait for more features like events etc.

Thank you so much for your work.

**Comment by netzih** [comment]
Not sure if to create a new issue or not.
But just noticed that when submitting a profile or contribution remotely it creates a new contact instead of using the regular CiviCrm duplicate rules... Is that a known issue for now?

Thank you

**Comment by gah242s** [comment]
Can I assume that you've already changed the setting on the Profile to tell it to use the dedupe feature rather than create a new record?

**Comment by netzih** [comment]
Yes.
When submitting the profile on my site it does not create a new contact.

**Comment by jmcclelland** [comment]
Glad the fix is working!

I've noticed the dupe bug too - would you mind opening a new ticket? I think it's a new bug. I suspect the bug is in core, but will have to dig in.
